contracts/Bidder.sol

```diff
@@ -16,32 +16,51 @@ contract Bidder is SimpleFlopper {
   );
 
   address public owner;
-  uint256 public bid;
+  uint256 public auctionID;
   uint256 public expectedLot;
+  bool public bidSubmited;
 
-  constructor(uint256 bid_) public {
+  /// @param auctionID_ ID of auction to participate in
+  constructor(address auctionAddress_, uint256 auctionID_) public SimpleFlopper(auctionAddress_) {
     owner = msg.sender;
-    bid = bid_;
+    auctionID = auctionID_;
   }
 
+  /// @notice Enters a given auction with bid of 1 mkr == 100 DAI
   function submitBid() external {
     require(msg.sender == owner, "Bidder/submitBid: owner only");
+    require(!bidSubmited, "Bidder/submitBid: bid alread bid submited");
 
     // dai has 45 decimal places
-    (uint256 amountDai, , , , ) = SimpleFlopper.getCurrentBid(bid);
+    (uint256 amountDai, , , , ) = SimpleFlopper.getCurrentBid(auctionID);
 
-    // lot needs to have 18 decimal places, and we're expecting 1 mkr = 100 dai
+    // lot needs to have 18 decimal places, and we're expecting 1 mkr == 100 dai
     expectedLot = (amountDai / 1e27) / 100;
 
     _VAT.move(owner, address(this), amountDai);
-    SimpleFlopper._bid(bid, expectedLot, amountDai);
+    // ??? Does our Bidder need to approve the Flopper?
+    SimpleFlopper._bid(auctionID, expectedLot, amountDai);
+
+    // Mark bid as successful 
+    bidSubmited = true;
   }
 
+  // Will withdraw funds to owner if auction is concluded or outbidded
   function finalize() external returns (bool) {
     require(msg.sender == owner, "Bidder/finalize: owner only");
-
-    ( , , , , uint48 end) = SimpleFlopper.getCurrentBid(bid);
-    require(end == 0, "Bidder/finalize: auction not finished");
+    require(bidSubmited, "Bidder/finalize: Bid not yet submitted");
+
+    // If auction was finalized, end should be 0x0.
+    ( , ,address bidder, , uint48 end) = SimpleFlopper.getCurrentBid(auctionID);
+
+    // If auction isn't closed, we try to close it ourselves
+    if (end != 0) {
+      // If we are the winning bidder, we finalize the auction
+      // Otherwise we got outbid and we withdraw DAI
+      if (bidder == address(this)) {
```

Comment:
Are we sure we don't need to finalize the auction in the event we lost?

Reply:
If someone outbid us, I believe they send back DAI to our Bidder right away as per https://github.com/makerdao/dss/blob/44330065999621834b08de1edf3b962f6bbd74c6/src/flop.sol#L136